composer/models/deeplabv3/model.py

```diff
@@ -186,14 +193,28 @@ def composer_deeplabv3(num_classes: int,
                       sync_bn=sync_bn,
                       initializers=initializers)
 
-    train_metrics = MetricCollection([CrossEntropy(ignore_index=-1), MIoU(num_classes, ignore_index=-1)])
-    val_metrics = MetricCollection([CrossEntropy(ignore_index=-1), MIoU(num_classes, ignore_index=-1)])
-
-    loss_fn = loss_registry['soft_cross_entropy']
-    loss_fn = functools.partial(loss_fn, ignore_index=-1)
+    train_metrics = MetricCollection(
+        [CrossEntropy(ignore_index=ignore_index),
+         MIoU(num_classes, ignore_index=ignore_index)])
+    val_metrics = MetricCollection(
+        [CrossEntropy(ignore_index=ignore_index),
+         MIoU(num_classes, ignore_index=ignore_index)])
+
+    ce_loss_fn = functools.partial(soft_cross_entropy, ignore_index=ignore_index)
+    dice_loss_fn = DiceLoss(softmax=True, batch=True, ignore_absent_classes=True)
+
+    def _combo_loss(output, target):
```

Comment:
curious why multiple losses are required here, versus a single loss:
```
loss_fn = lambda output, target: ce_loss_fn(output, target) * cross_entropy_weight + dice_loss_fn(output, target) * dice_weight
```

Reply:
With the lambda above, dice would be calculated even if the weight is 0, so we will get a 10% throughput hit for a few lines of code, did not seem worth it to me.

I think there are two benefits to returning a tuple of losses:
1. We should be plotting individual losses instead of total loss. This is a standard practice to measure optimization on the individual losses. We aren't doing this in the trainer yet, but we absolutely should.
2. There are a few potential algorithms that do automatic multiple loss reweighting. Although there is a lot of skepticism around these algorithms, it would be great if we could test them by implementing them as an `Algorithm`. If we sum the losses within the model, I imagine we would need to implement these algorithms within the model.

Reply:
Thanks for the explanation, very helpful!